framework/src/meshgenerators/MeshDiagnosticsGenerator.C

```diff
@@ -0,0 +1,1196 @@
+      const auto num_children = parent_ptr->n_children();
+
+      // Compare with the original set of elements
+      // We already know the child share the exterior node. If they share the same vertex
+      // average as the group of unrefined elements we will call this good enough for now
+      // For tetrahedral elements we cannot rely on the children all matching as the choice in
+      // the diagonal selection can be made differently. We'll just say 4 matching children is
+      // good enough for the heuristic
+      unsigned int num_children_match = 0;
+      for (const auto & child : parent_ptr->child_ref_range())
+      {
+        for (const auto & potential_children : fine_elements)
+          if (MooseUtils::absoluteFuzzyEqual(child.vertex_average()(0),
+                                             potential_children->vertex_average()(0),
+                                             _non_conformality_tol) &&
+              MooseUtils::absoluteFuzzyEqual(child.vertex_average()(1),
+                                             potential_children->vertex_average()(1),
+                                             _non_conformality_tol) &&
+              MooseUtils::absoluteFuzzyEqual(child.vertex_average()(2),
+                                             potential_children->vertex_average()(2),
+                                             _non_conformality_tol))
+          {
+            num_children_match++;
+            break;
+          }
+      }
+
+      if (num_children_match == num_children ||
+          ((elem_type == TET4 || elem_type == TET10 || elem_type == TET14) &&
+           num_children_match == 4))
+      {
+        num_likely_AMR_created_nonconformality++;
+        if (num_likely_AMR_created_nonconformality < 10)
+        {
+          _console << "Detected non-conformality likely created by AMR near" << *node
+                   << Moose::stringify(elem_type)
+                   << " elements that could be merged into a coarse element:" << std::endl;
+          for (const auto & elem : fine_elements)
+            _console << elem->id() << " ";
+          _console << std::endl;
+        }
+        else if (num_likely_AMR_created_nonconformality == 10)
+          _console << "Maximum log output reached, silencing output" << std::endl;
+      }
+    }
+
+    diagnosticsLog(
+        "Number of non-conformal nodes likely due to mesh refinement detected by heuristic: " +
+            Moose::stringify(num_likely_AMR_created_nonconformality),
+        _check_adaptivity_non_conformality,
+        num_likely_AMR_created_nonconformality);
+    pl->unset_close_to_point_tol();
+  }
+
+  if (_check_local_jacobian != "NO_CHECK")
+  {
+    unsigned int num_negative_elem_qp_jacobians = 0;
+    // Check elements (assumes serialized mesh)
+    for (const auto & elem : mesh->element_ptr_range())
+    {
+      // Get a high-ish order quadrature
+      QGauss qrule(mesh->mesh_dimension(), FOURTH);
```

Comment:
It wouldn't be crazy to use `FIFTH` order or user-selectable order or even loop over quadrature order here.  In a real simulation, even a quadratic FE type is likely to trigger a `FIFTH` order quadrature, so with `FOURTH` we're already potentially missing barely-singular elements that could fail in practice.  And although I know MOOSE users tend to stay low order, IIRC there's Cahn-Hilliard codes that end up with `SEVENTH` order quadrature to deal with weighted mass matrices with third-order bases.

Just a thought, though.  It's really unlikely to have a mesh that's messed up badly enough to have a singular element but not badly enough for fourth-order quadrature to catch it.  But it's also unlikely to have a mesh that's messed up in a way that an interior rule won't catch it but a side rule will, and you're (smartly, IMHO) also testing on sides to catch those.

Reply:
Oh!  And I'd pull this (and the FE, and the side FE, and the side quadrature rule) out of the element loop.  We can save some memory reallocation and often skip some reinitialization that way.

Reply:
fifth is fine

Reply:
I moved the quadrature and the FE type out of the loop. I guess this is not the part that matters

Reply:
ok it looks good actually now